### PR TITLE
Nsq search levels

### PR DESCRIPTION
**Pull request by NaderMahdi**
**Added support non-square block (nsq) search levels for better speed-quality trade-offs**

    // NSQ search Level                               Settings
    // 0                                                         OFF
    // 1                                                         Allow only NSQ Intra-FULL if parent SQ is intra-coded and vice versa.
    // 2                                                         Allow only NSQ Inter-NEAREST/NEAR/GLOBAL if parent SQ has no coeff
    // 3                                                         Allow only NSQ Intra-FULL and Inter-NEWMV if parent SQ is NEWMV
    // 4                                                         Allow only NSQ Inter-FULL and Intra-Z3 if parent SQ is intra-coded
    // 5                                                         Allow NSQ Intra-FULL and Inter-FULL

This feature is Off by default.

### CODE REVIEW

File: Source/Lib/Codec/EbCodingLoop.c [hunk 1]
```diff
@@ -630,12 +630,23 @@ static void Av1EncodeLoop(
             residual16bit->strideY,
             context_ptr->blk_geom->tx_width[context_ptr->txb_itr],
             context_ptr->blk_geom->tx_height[context_ptr->txb_itr]);
+#if TX_SEARCH_LEVELS
+        uint8_t  tx_search_skip_fag = picture_control_set_ptr->parent_pcs_ptr->tx_search_level == TX_SEARCH_ENC_DEC ? get_skip_tx_search_flag(
```
Comment (joelsole) [inline Source/Lib/Codec/EbCodingLoop.c]: tx_search_skip_fag => tx_search_skip_flag

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 2]
```diff
@@ -2001,17 +2003,34 @@ static void CflPrediction(
...
 
+#if TX_SEARCH_LEVELS
+uint8_t get_skip_tx_search_flag(
+    int32_t                  sq_size,
+    uint64_t                 ref_fast_cost,
+    uint64_t                 cu_cost,
+    uint64_t                 weight) 
+{
+
+    uint8_t  tx_search_skip_fag = cu_cost >= ((ref_fast_cost * weight) / 100) ? 1 : 0;
```
Comment (joelsole) [inline Source/Lib/Codec/EbProductCodingLoop.c]: Could you add a comment wrt this formula?

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 3]
```diff
@@ -2441,6 +2481,313 @@ EbBool allowed_ns_cu(
...
+
+        ProductFullLoopTxSearch(
+            candidateBuffer,
+            context_ptr,
+            picture_control_set_ptr
+        );
+
+
+        candidate_ptr->full_distortion = 0;
+
```
Comment (joelsole) [inline Source/Lib/Codec/EbProductCodingLoop.c]: too sparse

File: Source/Lib/Codec/EbDefinitions.h [hunk 4]
```diff
@@ -480,9 +476,33 @@ typedef struct InterpFilterParams {
...
+#if NSQ_SEARCH_LEVELS
+typedef enum NSQ_SEARCH_LEVEL {
+    NSQ_SEARCH_OFF,
+    NSQ_SEARCH_BASE_ON_SQ_TYPE,
+    NSQ_SEARCH_BASE_ON_SQ_COEFF,
+    NSQ_INTER_SEARCH_BASE_ON_SQ_MVMODE,
+    NSQ_INTER_SEARCH_BASE_ON_SQ_INTRAMODE,
+    NSQ_SEARCH_FULL
+} NSQ_SEARCH_LEVEL;
+#endif
```
Comment (mafonsoN) [inline Source/Lib/Codec/EbDefinitions.h]: Maybe missing indentation

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 5]
```diff
@@ -2441,6 +2481,313 @@ EbBool allowed_ns_cu(
...
+            context_ptr->md_local_cu_unit[cu_ptr->mds_idx].cost_luma = candidateBuffer->full_cost_luma;
+        }
+
+
+        context_ptr->md_ep_pipe_sb[cu_ptr->mds_idx].merge_cost = *candidateBuffer->full_cost_merge_ptr;
+        context_ptr->md_ep_pipe_sb[cu_ptr->mds_idx].skip_cost = *candidateBuffer->full_cost_skip_ptr;
+
+
+        if (candidate_ptr->type == INTER_MODE && candidate_ptr->merge_flag == EB_TRUE) {
+            context_ptr->md_ep_pipe_sb[cu_ptr->leaf_index].chroma_distortion = candidateBuffer->candidate_ptr->chroma_distortion;
```
Comment (joelsole) [inline Source/Lib/Codec/EbProductCodingLoop.c]: braces are only necessary for multi-line code blocks or multi-line condition statements;

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 6]
```diff
@@ -2441,6 +2481,313 @@ EbBool allowed_ns_cu(
...
+        candidateBuffer->cr_distortion[DIST_CALC_PREDICTION] = crFullDistortion[DIST_CALC_PREDICTION];
+        candidateBuffer->cr_coeff_bits = cr_coeff_bits;
+
+        candidateBuffer->candidate_ptr->full_distortion = (uint32_t)(y_full_distortion[0]);
+        candidateBuffer->candidate_ptr->luma_distortion = (uint32_t)(y_full_distortion[0]);
+
+
+        candidateBuffer->y_coeff_bits = y_coeff_bits;
+
+        if (picture_control_set_ptr->parent_pcs_ptr->chroma_mode == CHROMA_MODE_BEST)
```
Comment (joelsole) [inline Source/Lib/Codec/EbProductCodingLoop.c]: int my_func(const array *const values, const int num) {
    ..
}
braces go on the same line for single-line statements, but on a new line for multi-line statements

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 7]
```diff
@@ -2441,6 +2481,313 @@ EbBool allowed_ns_cu(
...
+                uint32_t j;
+
+                for (j = 0; j < bheight; j++)
+                {
+                    memcpy(dstPtr + j * bwidth, srcPtr + j * bwidth, bwidth * sizeof(int32_t));
+                }
+
+                if (context_ptr->blk_geom->has_uv)
+                {
+                    // Cb
```
Comment (joelsole) [inline Source/Lib/Codec/EbProductCodingLoop.c]: ditto

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 8]
```diff
@@ -2441,6 +2481,313 @@ EbBool allowed_ns_cu(
...
+                uint32_t  bwidth = context_ptr->blk_geom->tx_width[txb_itr] < 64 ? context_ptr->blk_geom->tx_width[txb_itr] : 32;
+                uint32_t  bheight = context_ptr->blk_geom->tx_height[txb_itr] < 64 ? context_ptr->blk_geom->tx_height[txb_itr] : 32;
+
+                int32_t* srcPtr = &(((int32_t*)buffer_ptr_array[lowestCostIndex]->residualQuantCoeffPtr->bufferY)[txb_1d_offset]);
+                int32_t* dstPtr = &(((int32_t*)context_ptr->cu_ptr->coeff_tmp->bufferY)[txb_1d_offset]);
+
+                uint32_t j;
+
+                for (j = 0; j < bheight; j++)
+                {
```
Comment (joelsole) [inline Source/Lib/Codec/EbProductCodingLoop.c]: no braces

File: Source/Lib/Codec/EbDefinitions.h [hunk 9]
```diff
@@ -480,9 +476,33 @@ typedef struct InterpFilterParams {
...
-
-
+#if INTERPOLATION_SEARCH_LEVELS
+typedef enum INTERPOLATION_SEARCH_LEVEL {
+    IT_SEARCH_OFF,
+    IT_SEARCH_INTER_DEPTH,
+    IT_SEARCH_FULL_LOOP,
+    IT_SEARCH_FAST_LOOP,
+} INTERPOLATION_SEARCH_LEVEL;
+#endif
```
Comment (mafonsoN) [inline Source/Lib/Codec/EbDefinitions.h]: maybe missing indentation

File: Source/Lib/Codec/EbRestorationPick.c [hunk 10]
```diff
@@ -709,11 +751,21 @@ static void search_sgrproj(const RestorationTileLimits *limits,
     const int32_t procunit_width = RESTORATION_PROC_UNIT_SIZE >> ss_x;
     const int32_t procunit_height = RESTORATION_PROC_UNIT_SIZE >> ss_y;
 
+#if FAST_SG
+    int8_t step = cm->sg_filter_mode == 1 ? 0 : cm->sg_filter_mode == 2 ? 1 : cm->sg_filter_mode == 3 ? 4 : 16;
```
Comment (joelsole) [inline Source/Lib/Codec/EbRestorationPick.c]: this is difficult to follow. Maybe a switch?

File: Source/Lib/Codec/EbProductCodingLoop.c (hunk 3 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: Addressed in previous commit.

File: Source/Lib/Codec/EbCodingLoop.c (hunk 1 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbCodingLoop.c]: Done

File: Source/Lib/Codec/EbProductCodingLoop.c (hunk 2 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: Done

File: Source/Lib/Codec/EbProductCodingLoop.c (hunk 5 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: Done

File: Source/Lib/Codec/EbProductCodingLoop.c (hunk 6 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: Done

File: Source/Lib/Codec/EbProductCodingLoop.c (hunk 8 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: Done

File: Source/Lib/Codec/EbDefinitions.h (hunk 9 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbDefinitions.h]: could you please provide more clarification?

File: Source/Lib/Codec/EbRestorationPick.c (hunk 10 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbRestorationPick.c]: Done

File: Source/Lib/Codec/EbProductCodingLoop.c (hunk 7 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbProductCodingLoop.c]: Done

**Pull request closed without merging**